float64 output_x
float64 output_y
bool valid